apps/ewallet_db/priv/repo/migrations/20190124074130_cleanup_hierarchy.exs

```diff
@@ -0,0 +1,287 @@
+    end
+
+    create unique_index(:membership, [:key_uuid, :account_uuid])
+
+    alter table(:user) do
+      add :global_role, :string
+    end
+
+    alter table(:key) do
+      add :global_role, :string
+      remove :account_uuid
+    end
+
+    flush()
+
+    case up_get_master_account_uuid() do
+      nil -> nil
+      [master_uuid, master_id] ->
+        ensure_master_account_setting()
+
+        # give super admin to admins of master account
+        give_super_admin_to_master_account_admins(master_uuid)
+        give_super_admin_to_all_keys()
+
+        # set master account in settings
+        set_master_account_in_settings(master_id)
+
+        # Add memberships for all children accounts
+        add_memberships_for_all_children_accounts(master_uuid)
+
+        # global roles for end users and admins
+        set_end_users_global_role()
+    end
+
+    alter table(:account) do
+      remove :parent_uuid
+    end
+  end
+
+  def down do
+    alter table(:account) do
+      add :parent_uuid, references(:account, column: :uuid, type: :uuid)
+    end
+
+    alter table(:key) do
+      add :account_uuid, references(:account, column: :uuid, type: :uuid)
+    end
+
+    flush()
+
+    case down_get_master_account_uuid() do
+      [master_uuid, _master_id] ->
+        # set all account as children of master account (except master, duh!)
+        set_all_account_as_master_children(master_uuid)
+
+        # move super admins to master account admin memberships
+        add_super_admins_to_master_account(master_uuid)
+
+        # Add master account to keys
+        add_master_account_to_keys(master_uuid)
+
+        # remove master account from settings
+        remove_master_account_from_settings()
+
+        # remove children memberships
+        remove_children_memberships(master_uuid)
+      _ ->
+        nil
+    end
+
+    alter table(:membership) do
+      remove :key_uuid
+    end
+
+    alter table(:user) do
+      remove :global_role, :string
+    end
+
+    alter table(:key) do
+      remove :global_role, :string
+    end
+  end
+
+  # This is only ran if the DB already has data, which requires a settings seeding to proceed.
+  # Clean DBs won't be impacted.
+  defp ensure_master_account_setting do
+
+    if Repo.one(from(s in "setting", select: s.uuid, where: s.key == "master_account")) == nil do
+      raise "Run seed for settings before this migration with 'mix seed --settings'"
+    end
+  end
+
+  defp up_get_master_account_uuid do
+    Repo.one(from(a in "account",
+             select: [a.uuid, a.id],
+             where: is_nil(a.parent_uuid)))
+  end
+
+  defp down_get_master_account_uuid do
+    data = Repo.one(from(s in "setting",
+                        select: s.data,
+                        where: s.key == "master_account"))
+    case data["value"] do
+      nil ->
+        Repo.one(from a in "account", select: [a.uuid, a.id], order_by: [asc: a.inserted_at], limit: 1)
```

Comment:
Noted that this will assume the first inserted account becomes master account if it cannot be found in the settings.